SearchAsyncActionTests.testFanOutAndCollect throws RejectedExecutionException

I dug into #29242 again and tried to reproduce the problem at 5a3fd8e4e7c8d2721b3542ffa859c9245678d6c2 by running the affected test repeatedly until it failed by using:

    date | tee pass-times.log; while ./gradlew --console=plain :server:test -Dtests.class=org.elasticsearch.action.search.SearchAsyncActionTests -Dtests.method="testFanOutAndCollect" 2>&1 > testoutput.log; do date | tee -a pass-times.log; done

After around 1000 iterations (6 hours) it failed with a different issue:

```
  2> ༩ ༡༣, ༢༠༡༨ ༢:༤༧:༢༧ ཕྱི་ཆ་ com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
  2> WARNING: Uncaught exception in thread: Thread[Thread-1,5,TGRP-SearchAsyncActionTests]
  2> java.util.concurrent.RejectedExecutionException: Task org.elasticsearch.action.search.InitialSearchPhase$1@db8dd7d rejected from java.util.concurrent.ThreadPoolExecutor@2e3a06e7[Terminated,
 pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
  2>    at __randomizedtesting.SeedInfo.seed([CAEABDE40791E9D5]:0)
  2>    at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2080)
  2>    at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:832)
  2>    at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1365)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.fork(InitialSearchPhase.java:174)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:167)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.executeNext(InitialSearchPhase.java:239)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.lambda$performPhaseOnShard$4(InitialSearchPhase.java:293)
  2>    at org.elasticsearch.action.search.InitialSearchPhase$PendingExecutions.tryRun(InitialSearchPhase.java:214)
  2>    at org.elasticsearch.action.search.InitialSearchPhase$PendingExecutions.finishAndRunNext(InitialSearchPhase.java:208)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:169)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.executeNext(InitialSearchPhase.java:239)
  2>    at org.elasticsearch.action.search.InitialSearchPhase.access$100(InitialSearchPhase.java:50)
  2>    at org.elasticsearch.action.search.InitialSearchPhase$2.innerOnResponse(InitialSearchPhase.java:270)
  2>    at org.elasticsearch.action.search.SearchActionListener.onResponse(SearchActionListener.java:45)
  2>    at org.elasticsearch.action.search.SearchAsyncActionTests$7.lambda$executePhaseOnShard$1(SearchAsyncActionTests.java:334)
  2>    at java.base/java.lang.Thread.run(Thread.java:844)
  2> REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=CAEABDE40791E9D5 -Dtests.class=org.elasticsearch.action.search.SearchAsyncActionTests -Dtests.method="testFanOutAndCollect" -Dtests.security.manager=true -Dtests.locale=dz -Dtests.timezone=Etc/GMT+11 -Dcompiler.java=10 -Druntime.java=10
ERROR   0.14s | SearchAsyncActionTests.testFanOutAndCollect <<< FAILURES!
   > Throwable #1: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=18, name=Thread-1, state=RUNNABLE, group=TGRP-SearchAsyncActionTests]
   >    at __randomizedtesting.SeedInfo.seed([CAEABDE40791E9D5:B11B3894FE3A2856]:0)
   > Caused by: java.util.concurrent.RejectedExecutionException: Task org.elasticsearch.action.search.InitialSearchPhase$1@db8dd7d rejected from java.util.concurrent.ThreadPoolExecutor@2e3a06e7[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
   >    at __randomizedtesting.SeedInfo.seed([CAEABDE40791E9D5]:0)
   >    at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2080)
   >    at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:832)
   >    at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1365)
   >    at org.elasticsearch.action.search.InitialSearchPhase.fork(InitialSearchPhase.java:174)
   >    at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:167)
   >    at org.elasticsearch.action.search.InitialSearchPhase.executeNext(InitialSearchPhase.java:239)
   >    at org.elasticsearch.action.search.InitialSearchPhase.lambda$performPhaseOnShard$4(InitialSearchPhase.java:293)
   >    at org.elasticsearch.action.search.InitialSearchPhase$PendingExecutions.tryRun(InitialSearchPhase.java:214)
   >    at org.elasticsearch.action.search.InitialSearchPhase$PendingExecutions.finishAndRunNext(InitialSearchPhase.java:208)
   >    at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:169)
   >    at org.elasticsearch.action.search.InitialSearchPhase.executeNext(InitialSearchPhase.java:239)
   >    at org.elasticsearch.action.search.InitialSearchPhase.access$100(InitialSearchPhase.java:50)
   >    at org.elasticsearch.action.search.InitialSearchPhase$2.innerOnResponse(InitialSearchPhase.java:270)
   >    at org.elasticsearch.action.search.SearchActionListener.onResponse(SearchActionListener.java:45)
   >    at org.elasticsearch.action.search.SearchAsyncActionTests$7.lambda$executePhaseOnShard$1(SearchAsyncActionTests.java:334)
   >    at java.base/java.lang.Thread.run(Thread.java:844)
  2> NOTE: test params are: codec=Asserting(Lucene80): {}, docValues:{}, maxPointsInLeafNode=1617, maxMBSortInHeap=7.294998490965383, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@5ee576f6), locale=dz, timezone=Etc/GMT+11
  2> NOTE: Linux 4.13.0-32-generic amd64/Oracle Corporation 10 (64-bit)/cpus=8,threads=1,free=494096088,total=536870912
  2> NOTE: All tests run in this JVM: [SearchAsyncActionTests]

FAILURE: Build failed with an exception.
```

I initially saw this kind of failure using `-Dtests.iters=1000`, which yielded this failure after only a few runs, but was unsure whether it might have been a symptom of the `tests.iters` setting. This confirms that it can also occur if the test runs individually.

--- a/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java
+++ b/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java
@@ -166,7 +166,6 @@ abstract class AbstractSearchAsyncAction<Result extends SearchPhaseResult> exten
         }
     }
 
-
     private ShardSearchFailure[] buildShardFailures() {
         AtomicArray<ShardSearchFailure> shardFailures = this.shardFailures.get();
         if (shardFailures == null) {

--- a/server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java
+++ b/server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java
@@ -160,14 +160,6 @@ abstract class InitialSearchPhase<FirstResult extends SearchPhaseResult> extends
         }
     }
 
-    private void maybeFork(final Thread thread, final Runnable runnable) {
-        if (thread == Thread.currentThread()) {
-            fork(runnable);
-        } else {
-            runnable.run();
-        }
-    }
-
     private void fork(final Runnable runnable) {
         executor.execute(new AbstractRunnable() {
             @Override
@@ -232,10 +224,18 @@ abstract class InitialSearchPhase<FirstResult extends SearchPhaseResult> extends
     }
 
     private void executeNext(PendingExecutions pendingExecutions, Thread originalThread) {
+        executeNext(pendingExecutions == null ? null : pendingExecutions::finishAndRunNext, originalThread);
+    }
+
+    protected void executeNext(Runnable runnable, Thread originalThread) {
         if (throttleConcurrentRequests) {
-            maybeFork(originalThread, pendingExecutions::finishAndRunNext);
+            if (originalThread == Thread.currentThread()) {
+                fork(runnable);
+            } else {
+                runnable.run();
+            }
         } else {
-            assert pendingExecutions == null;
+            assert runnable == null;
         }
     }
 
@@ -256,28 +256,26 @@ abstract class InitialSearchPhase<FirstResult extends SearchPhaseResult> extends
             Runnable r = () -> {
                 final Thread thread = Thread.currentThread();
                 try {
-                    executePhaseOnShard(shardIt, shard, new SearchActionListener<FirstResult>(
-                        shardIt.newSearchShardTarget(shard.currentNodeId()), shardIndex) {
-                        @Override
-                        public void innerOnResponse(FirstResult result) {
-                            try {
-                                onShardResult(result, shardIt);
-                            } finally {
-                                executeNext(pendingExecutions, thread);
+                    executePhaseOnShard(shardIt, shard,
+                        new SearchActionListener<FirstResult>(shardIt.newSearchShardTarget(shard.currentNodeId()), shardIndex) {
+                            @Override
+                            public void innerOnResponse(FirstResult result) {
+                                try {
+                                    onShardResult(result, shardIt);
+                                } finally {
+                                    executeNext(pendingExecutions, thread);
+                                }
                             }
-                        }
 
-                        @Override
-                        public void onFailure(Exception t) {
-                            try {
-                                onShardFailure(shardIndex, shard, shard.currentNodeId(), shardIt, t);
-                            } finally {
-                                executeNext(pendingExecutions, thread);
+                            @Override
+                            public void onFailure(Exception t) {
+                                try {
+                                    onShardFailure(shardIndex, shard, shard.currentNodeId(), shardIt, t);
+                                } finally {
+                                    executeNext(pendingExecutions, thread);
+                                }
                             }
-                        }
-                    });
-
-
+                        });
                 } catch (final Exception e) {
                     try {
                         /*

--- a/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
@@ -63,7 +63,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         SearchRequest request = new SearchRequest();
         request.allowPartialSearchResults(true);
         int numShards = 10;
-        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {}, e -> fail(e.getMessage()));
+        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {},
+            (e) -> { throw new AssertionError("unexpected", e);});
         DiscoveryNode primaryNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
         DiscoveryNode replicaNode = new DiscoveryNode("node_2", buildNewFakeTransportAddress(), Version.CURRENT);
 
@@ -161,7 +162,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         int numShards = 10;
         CountDownLatch latch = new CountDownLatch(numShards);
         AtomicBoolean searchPhaseDidRun = new AtomicBoolean(false);
-        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {}, e -> fail(e.getMessage()));
+        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {},
+            (e) -> { throw new AssertionError("unexpected", e);});
         DiscoveryNode primaryNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
         // for the sake of this test we place the replica on the same node. ie. this is not a mistake since we limit per node now
         DiscoveryNode replicaNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
@@ -258,7 +260,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         request.setMaxConcurrentShardRequests(randomIntBetween(1, 100));
         AtomicReference<TestSearchResponse> response = new AtomicReference<>();
         ActionListener<SearchResponse> responseListener = ActionListener.wrap(
-            searchResponse -> response.set((TestSearchResponse) searchResponse), e -> fail(e.getMessage()));
+            searchResponse -> response.set((TestSearchResponse) searchResponse),
+            (e) -> { throw new AssertionError("unexpected", e);});
         DiscoveryNode primaryNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
         DiscoveryNode replicaNode = new DiscoveryNode("node_2", buildNewFakeTransportAddress(), Version.CURRENT);
 

--- a/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
@@ -50,7 +50,6 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
@@ -62,7 +61,8 @@ public class SearchAsyncActionTests extends ESTestCase {
     public void testSkipSearchShards() throws InterruptedException {
         SearchRequest request = new SearchRequest();
         request.allowPartialSearchResults(true);
-        CountDownLatch latch = new CountDownLatch(1);
+        int numShards = 10;
+        CountDownLatch latch = new CountDownLatch(numShards);
         AtomicReference<TestSearchResponse> response = new AtomicReference<>();
         ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
             @Override
@@ -82,7 +82,7 @@ public class SearchAsyncActionTests extends ESTestCase {
         AtomicInteger contextIdGenerator = new AtomicInteger(0);
         GroupShardsIterator<SearchShardIterator> shardsIter = getShardsIter("idx",
             new OriginalIndices(new String[]{"idx"}, SearchRequest.DEFAULT_INDICES_OPTIONS),
-            10, randomBoolean(), primaryNode, replicaNode);
+            numShards, randomBoolean(), primaryNode, replicaNode);
         int numSkipped = 0;
         for (SearchShardIterator iter : shardsIter) {
             if (iter.shardId().id() % 2 == 0) {
@@ -142,9 +142,15 @@ public class SearchAsyncActionTests extends ESTestCase {
                     return new SearchPhase("test") {
                         @Override
                         public void run() {
-                            latch.countDown();
+
                         }
                     };
+                }
+
+                @Override
+                protected void executeNext(Runnable runnable, Thread originalThread) {
+                    super.executeNext(runnable, originalThread);
+                    latch.countDown();
                 }
             };
         asyncAction.start();
@@ -161,7 +167,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         request.allowPartialSearchResults(true);
         int numConcurrent = randomIntBetween(1, 5);
         request.setMaxConcurrentShardRequests(numConcurrent);
-        CountDownLatch latch = new CountDownLatch(1);
+        int numShards = 10;
+        CountDownLatch latch = new CountDownLatch(numShards);
         AtomicReference<TestSearchResponse> response = new AtomicReference<>();
         ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
             @Override
@@ -182,7 +189,7 @@ public class SearchAsyncActionTests extends ESTestCase {
         AtomicInteger contextIdGenerator = new AtomicInteger(0);
         GroupShardsIterator<SearchShardIterator> shardsIter = getShardsIter("idx",
             new OriginalIndices(new String[]{"idx"}, SearchRequest.DEFAULT_INDICES_OPTIONS),
-            10, randomBoolean(), primaryNode, replicaNode);
+            numShards, randomBoolean(), primaryNode, replicaNode);
         SearchTransportService transportService = new SearchTransportService(null, null);
         Map<String, Transport.Connection> lookup = new HashMap<>();
         Map<ShardId, Boolean> seenShard = new ConcurrentHashMap<>();
@@ -246,9 +253,15 @@ public class SearchAsyncActionTests extends ESTestCase {
                     return new SearchPhase("test") {
                         @Override
                         public void run() {
-                            latch.countDown();
+
                         }
                     };
+                }
+
+                @Override
+                protected void executeNext(Runnable runnable, Thread originalThread) {
+                    super.executeNext(runnable, originalThread);
+                    latch.countDown();
                 }
             };
         asyncAction.start();
@@ -280,9 +293,10 @@ public class SearchAsyncActionTests extends ESTestCase {
 
         Map<DiscoveryNode, Set<Long>> nodeToContextMap = newConcurrentMap();
         AtomicInteger contextIdGenerator = new AtomicInteger(0);
+        int numShards = randomIntBetween(1, 10);
         GroupShardsIterator<SearchShardIterator> shardsIter = getShardsIter("idx",
                 new OriginalIndices(new String[]{"idx"}, SearchRequest.DEFAULT_INDICES_OPTIONS),
-                randomIntBetween(1, 10), randomBoolean(), primaryNode, replicaNode);
+                numShards, randomBoolean(), primaryNode, replicaNode);
         AtomicInteger numFreedContext = new AtomicInteger();
         SearchTransportService transportService = new SearchTransportService(null, null) {
             @Override
@@ -296,9 +310,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         lookup.put(primaryNode.getId(), new MockConnection(primaryNode));
         lookup.put(replicaNode.getId(), new MockConnection(replicaNode));
         Map<String, AliasFilter> aliasFilters = Collections.singletonMap("_na_", new AliasFilter(null, Strings.EMPTY_ARRAY));
-        final ExecutorService executor = Executors.newFixedThreadPool(randomIntBetween(1, Runtime.getRuntime().availableProcessors()));
-        final CountDownLatch latch = new CountDownLatch(1);
-        final AtomicBoolean latchTriggered = new AtomicBoolean();
+        ExecutorService executor = Executors.newFixedThreadPool(randomIntBetween(1, Runtime.getRuntime().availableProcessors()));
+        final CountDownLatch latch = new CountDownLatch(numShards);
         AbstractSearchAsyncAction<TestSearchPhaseResult> asyncAction =
                 new AbstractSearchAsyncAction<TestSearchPhaseResult>(
                         "test",
@@ -349,12 +362,14 @@ public class SearchAsyncActionTests extends ESTestCase {
                             sendReleaseSearchContext(result.getRequestId(), new MockConnection(result.node), OriginalIndices.NONE);
                         }
                         responseListener.onResponse(response);
-                        if (latchTriggered.compareAndSet(false, true) == false) {
-                            throw new AssertionError("latch triggered twice");
-                        }
-                        latch.countDown();
                     }
                 };
+            }
+
+            @Override
+            protected void executeNext(Runnable runnable, Thread originalThread) {
+                super.executeNext(runnable, originalThread);
+                latch.countDown();
             }
         };
         asyncAction.start();

--- a/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
@@ -50,6 +50,7 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
@@ -62,19 +63,7 @@ public class SearchAsyncActionTests extends ESTestCase {
         SearchRequest request = new SearchRequest();
         request.allowPartialSearchResults(true);
         int numShards = 10;
-        AtomicReference<TestSearchResponse> response = new AtomicReference<>();
-        ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
-            @Override
-            public void onResponse(SearchResponse searchResponse) {
-                response.set((TestSearchResponse) searchResponse);
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                logger.warn("test failed", e);
-                fail(e.getMessage());
-            }
-        };
+        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {}, e -> fail(e.getMessage()));
         DiscoveryNode primaryNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
         DiscoveryNode replicaNode = new DiscoveryNode("node_2", buildNewFakeTransportAddress(), Version.CURRENT);
 
@@ -90,6 +79,7 @@ public class SearchAsyncActionTests extends ESTestCase {
             }
         }
         CountDownLatch latch = new CountDownLatch(numShards - numSkipped);
+        AtomicBoolean searchPhaseDidRun = new AtomicBoolean(false);
 
         SearchTransportService transportService = new SearchTransportService(null, null);
         Map<String, Transport.Connection> lookup = new HashMap<>();
@@ -142,7 +132,7 @@ public class SearchAsyncActionTests extends ESTestCase {
                     return new SearchPhase("test") {
                         @Override
                         public void run() {
-
+                            assertTrue(searchPhaseDidRun.compareAndSet(false, true));
                         }
                     };
                 }
@@ -155,6 +145,7 @@ public class SearchAsyncActionTests extends ESTestCase {
             };
         asyncAction.start();
         latch.await();
+        assertTrue(searchPhaseDidRun.get());
         SearchResponse searchResponse = asyncAction.buildSearchResponse(null, null);
         assertEquals(shardsIter.size() - numSkipped, numRequests.get());
         assertEquals(0, searchResponse.getFailedShards());
@@ -169,19 +160,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         request.setMaxConcurrentShardRequests(numConcurrent);
         int numShards = 10;
         CountDownLatch latch = new CountDownLatch(numShards);
-        AtomicReference<TestSearchResponse> response = new AtomicReference<>();
-        ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
-            @Override
-            public void onResponse(SearchResponse searchResponse) {
-                response.set((TestSearchResponse) searchResponse);
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                logger.warn("test failed", e);
-                fail(e.getMessage());
-            }
-        };
+        AtomicBoolean searchPhaseDidRun = new AtomicBoolean(false);
+        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {}, e -> fail(e.getMessage()));
         DiscoveryNode primaryNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
         // for the sake of this test we place the replica on the same node. ie. this is not a mistake since we limit per node now
         DiscoveryNode replicaNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
@@ -253,7 +233,7 @@ public class SearchAsyncActionTests extends ESTestCase {
                     return new SearchPhase("test") {
                         @Override
                         public void run() {
-
+                            assertTrue(searchPhaseDidRun.compareAndSet(false, true));
                         }
                     };
                 }
@@ -268,6 +248,7 @@ public class SearchAsyncActionTests extends ESTestCase {
         assertEquals(numConcurrent, numRequests.get());
         awaitInitialRequests.countDown();
         latch.await();
+        assertTrue(searchPhaseDidRun.get());
         assertEquals(10, numRequests.get());
     }
 
@@ -276,18 +257,8 @@ public class SearchAsyncActionTests extends ESTestCase {
         request.allowPartialSearchResults(true);
         request.setMaxConcurrentShardRequests(randomIntBetween(1, 100));
         AtomicReference<TestSearchResponse> response = new AtomicReference<>();
-        ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
-            @Override
-            public void onResponse(SearchResponse searchResponse) {
-                response.set((TestSearchResponse) searchResponse);
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                logger.warn("test failed", e);
-                fail(e.getMessage());
-            }
-        };
+        ActionListener<SearchResponse> responseListener = ActionListener.wrap(
+            searchResponse -> response.set((TestSearchResponse) searchResponse), e -> fail(e.getMessage()));
         DiscoveryNode primaryNode = new DiscoveryNode("node_1", buildNewFakeTransportAddress(), Version.CURRENT);
         DiscoveryNode replicaNode = new DiscoveryNode("node_2", buildNewFakeTransportAddress(), Version.CURRENT);
 

--- a/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java
@@ -62,7 +62,6 @@ public class SearchAsyncActionTests extends ESTestCase {
         SearchRequest request = new SearchRequest();
         request.allowPartialSearchResults(true);
         int numShards = 10;
-        CountDownLatch latch = new CountDownLatch(numShards);
         AtomicReference<TestSearchResponse> response = new AtomicReference<>();
         ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
             @Override
@@ -90,6 +89,7 @@ public class SearchAsyncActionTests extends ESTestCase {
                 numSkipped++;
             }
         }
+        CountDownLatch latch = new CountDownLatch(numShards - numSkipped);
 
         SearchTransportService transportService = new SearchTransportService(null, null);
         Map<String, Transport.Connection> lookup = new HashMap<>();
@@ -156,7 +156,7 @@ public class SearchAsyncActionTests extends ESTestCase {
         asyncAction.start();
         latch.await();
         SearchResponse searchResponse = asyncAction.buildSearchResponse(null, null);
-        assertEquals(shardsIter.size()-numSkipped, numRequests.get());
+        assertEquals(shardsIter.size() - numSkipped, numRequests.get());
         assertEquals(0, searchResponse.getFailedShards());
         assertEquals(numSkipped, searchResponse.getSkippedShards());
         assertEquals(shardsIter.size(), searchResponse.getSuccessfulShards());

